Asterix/optics/propagation_functions.py

```diff
@@ -610,30 +615,81 @@ def prop_fpm_regional_sampling(pup,
     EF_before_LS : 2D array (complex)
         E-field before the Lyot stop.
     """
-    from Asterix.optics import shift_phase_ramp
+    samp_outer = 2
+
+    if not isinstance(nbres, (list, np.ndarray)):
+        raise TypeError(f"'nbres' parameter need to be of type list or numpy.array. Currently type = {type(nbres)}")
 
-    if samp_outer != 2:
-        raise ValueError(f"samp_outer' needs to be 2, otherwise we cut off the high-spatial frequencies and the"
-                         f"simulation turns out bad. This can become a variable parameter in the future if we allow"
-                         f"the array sizes to be adapted in this function as well, in order to retain the original"
-                         f"spatial frequency content in the outer sampling layer.")
+    nbres = np.array(nbres)
 
-    dim_pup = pup.shape[0]
+    dim_overpad_pupil = pup.shape[0]
     dim_fpm = fpm.shape[0]
-    pup0 = np.array(pup, copy=True, dtype='complex128')
+
+    if real_dim_input is None:
+        real_dim_input = pup.shape[0]
+    samplings = real_dim_input / nbres
+
+    if not np.all(np.diff(nbres) >= 0):  # check if it is sorted by checking 2 by 2 diff is always positive
+        raise ValueError(f"'nbres' parameter need to be sorted from the highest to lowest nbrs of elements."
+                         f"Currently 'nbres' = {nbres}")
+
+    if np.min(samplings) < 2:
+        raise ValueError(f"The outer sampling in prop_fpm_regional_sampling is hardcoded 2, otherwise we cut off"
+                         f"the high-spatial frequencies and the simulation turns out bad. We need the samplings"
+                         f"defined by the 'nbres' parameter (real_dim_input/nbres) to be always >= 2. Currently, with"
+                         f"real_dim_input = {real_dim_input}, samplings are {samplings}")
+
+    if np.min(samplings) == 2:
+        # If the outer sampling defined by nbrs is already 2, we can remove it and gain some time
+        # because the last sampling is harcoded at 2
+        nbres = np.delete(nbres, -1)
+        samplings = np.delete(samplings, -1)
+
+    if max(shift) >= nbres[0] / 2:
+        raise ValueError(f"shift {shift} is larger than the minimum nbrs {nbres[0]} of element of resolution : the "
+                         f"tip/tilt is out of the array! Increase min(nbres) or decrease shift.")
+
+    # can be used to check:
+    # print(f"With real_dim_input = {real_dim_input} and nbrs = {nbres}, Samplings: ", samplings)
```

Comment:
Please delete. Commented print statements are very bad coding. If you need it in the future you can add it temporarily, otherwise it just clutters the code.